core/filterconstraint.cpp

```diff
@@ -0,0 +1,939 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "filterconstraint.h"
+#include "dive.h"
+#include "divesite.h"
+#include "errorhelper.h"
+#include "gettextfromc.h"
+#include "qthelper.h"
+#include "tag.h"
+#include "trip.h"
+#include "subsurface-time.h"
+#include <QDateTime>
+
+// We use the units enum only internally.
+// Therefore define it here, not in the header file.
+enum filter_constraint_units {
+	FILTER_CONSTRAINT_NO_UNIT = 0,
+	FILTER_CONSTRAINT_LENGTH_UNIT,
+	FILTER_CONSTRAINT_TEMPERATURE_UNIT,
+	FILTER_CONSTRAINT_WEIGHT_UNIT,
+	FILTER_CONSTRAINT_VOLUMETRIC_FLOW_UNIT
+};
+
+static struct type_description {
+	filter_constraint_type type;
+	const char *s;			// untranslated string that, which will be written to the log
+	const char *translated;
+	bool has_string_mode;
+	bool has_range_mode;
+	bool is_star_widget;
+	filter_constraint_units units;
+	int decimal_places;
+	bool has_date;
+	bool has_time;
+} type_descriptions[] = {
+	{ FILTER_CONSTRAINT_DATE, "date", QT_TRANSLATE_NOOP("gettextFromC", "date"), false, true, false, FILTER_CONSTRAINT_NO_UNIT, 0, true, false },
```

Comment:
so above you say that the third element is the _translated_ string, yet here you only mark it for translation but don't actually translate it as you initialize the array... I must be missing something...

Reply:
Yes, the name was unfortunate. I renamed it to `text_ui`.

Note that this follows a consistent pattern of our C code. See for example `divemode_text_ui` code. We can't translate it directly, because these functions will be run before `main()` runs. To my knowledge in C you couldn't even run functions for static initialization.

The NOOP macro is just supposed to add these strings to Qt's translation system. The translation is then done on-the-fly. Yes, we could translate the strings once at application startup, but that makes things more complicated (order-of-initialization) without any apparent gain.

In conclusion, apart from the silly name, I think this is fine and consistent with the C code.

Reply:
makes sense. Thanks